File: src/main/java/fm/comunas/ddd/property/model/base/PropertyBase.java
package fm.comunas.ddd.property.model.base;

import fm.comunas.ddd.property.model.Property;

public abstract class PropertyBase<T> implements Property<T> {

	private final EntityWithPropertiesSPI entity;
	private boolean isWritable = true;

	protected PropertyBase(EntityWithPropertiesSPI entity) {
		this.entity = entity;
	}

	protected EntityWithPropertiesSPI getEntity() {
		return this.entity;
	}

	@Override
	public abstract String getName();

	@Override
	public boolean isWritable() {
		return this.isWritable;
	}

	@Override
	public void setWritable(boolean isWritable) {
		this.isWritable = isWritable;
	}

}

File: fm-dddrive/src/main/java/io/dddrive/property/model/base/PropertyBase.kt
package io.dddrive.property.model.base

import io.dddrive.property.model.EntityWithProperties
import io.dddrive.property.model.EntityWithPropertiesSPI
import io.dddrive.property.model.Property

abstract class PropertyBase<T : Any>(
	override val entity: EntityWithProperties,
	override val name: String,
) : Property<T> {

	override val relativePath: String
		get() {
			val relativePath = (this.entity as EntityWithPropertiesSPI).relativePath
			return if (relativePath.isEmpty()) this.name else relativePath + "." + this.name
		}

	override val path: String
		get() = (this.entity as EntityWithPropertiesSPI).path + "." + this.name

	override val isWritable: Boolean
		get() = !this.entity.isFrozen

}
